### PR TITLE
Port Datasource

### PR DESCRIPTION
**Pull request by doriac11**
# Description
Implement the Port Datasource

# ISSUE TYPE
- Feature Pull Request

##### RESOURCE OR DATASOURCE NAME
Port Datasource

##### Example Response
```
  + portFilter = {
      + filter       = null
      + id           = "port-datasource"
      + port_details = [
          + {
              + aclx                                        = true
              + avoid_reset_broadcast                       = false
              + capable_protocol                            = [
                  + "SCSI_FC",
                ]
              + common_serial_number                        = true
              + director_id                                 = "OR-1C"
              + director_status                             = "Online"
              + disable_q_reset_on_ua                       = false
              + enable_auto_negotiate                       = true
              + enabled_protocol                            = [
                  + "SCSI_FC",
                ]
              + environ_set                                 = false
              + hp_3000_mode                                = false
              + identifier                                  = "5000097200097802"
              + init_point_to_point                         = true
              + ip_addresses                                = null
              + ipv4_address                                = null
              + ipv4_default_gateway                        = null
              + ipv4_domain_name                            = null
              + ipv4_netmask                                = null
              + ipv6_address                                = null
              + ipv6_prefix                                 = null
              + iscsi_endpoint                              = false
              + iscsi_target                                = false
              + mac_address                                 = null
              + maskingview                                 = [
                  + "pmax606_esx198115_mv_1",
                  + "terraform-pmax606-fc-ds-20t_MV",
                  + "csm-pmax606-fc-ds-3t_MV",
                  + "OS-pie105070-F-DER_PGc575a3-MV",
                  + "pmax606_esx198109_mv_1",
                  + "VMAX_Auto_1_MV",
                  + "VMAX_Auto_2_MV",
                  + "pmax606-fc-ds-3t_MV",
                  + "qzt0asc3p7m328tq2g5aarfgp7v0jr9f17pzcp0e",
                  + "tfacc_host_group_example_1_GK_MV",
                  + "VMAX_Parent_MV",
                ]
              + max_speed                                   = "32"
              + negotiate_reset                             = false
              + negotiated_speed                            = "16"
              + network_id                                  = null
              + no_participating                            = false
              + num_of_cores                                = null
              + num_of_hypers                               = null
              + num_of_mapped_vols                          = 539
              + num_of_masking_views                        = 11
              + num_of_port_groups                          = 9
              + nvmetcp_endpoint                            = false
              + port_id                                     = "2"
              + port_interface                              = null
              + port_status                                 = "ON"
              + portgroup                                   = [
                  + "csm-pmax606-fc-ds-3t_PG",
                  + "FC_OpenStack_PG",
                  + "PIE_IAPI_Automation_PG",
                  + "PIE_UI_Automation_PG",
                  + "pmax606-fc-ds-3t_PG",
                  + "PMAX_CINDER_PG",
                  + "terraform-pmax606-fc-ds-20t_PG",
                  + "tfacc_host_group_example_1_GK_PG",
                  + "VSI_SECURITY_TESTING_PG",
                ]
              + power_levels_last_sampled_date_milliseconds = 1689167517000
              + prevent_automatic_rdf_link_recovery         = null
              + prevent_ra_online_on_power_up               = null
              + rdf_hardware_compression                    = null
              + rdf_hardware_compression_supported          = null
              + rdf_ra_group_attributes_farpoint            = null
              + rdf_software_compression                    = null
              + rdf_software_compression_supported          = null
              + rx_power_level_mw                           = null
              + scsi_3                                      = true
              + scsi_support1                               = true
              + siemens                                     = false
              + soft_reset                                  = false
              + spc2_protocol_version                       = true
              + sunapee                                     = false
              + tcp_port                                    = null
              + tx_power_level_mw                           = null
              + type                                        = "OSHostAndRDF"
              + unique_wwn                                  = true
              + vcm_state                                   = "Enabled"
              + vnx_attached                                = false
              + volume_set_addressing                       = false
              + wwn_node                                    = "5000097200097bff"
              + z_hyperlink_port                            = null
            },
        ]
    }

```
# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 80% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Unit tests
- [x] Acceptance tests

### CODE REVIEW

**Comment by terraform-providers-svc** [comment]
Can one of the admins verify this patch?

File: powermax/helper/port_helper.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	"dell/powermax-go-client"
+	"fmt"
+	"strings"
+	"terraform-provider-powermax/client"
+	"terraform-provider-powermax/powermax/models"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// FilterPortIds Based on state either use the filtered list of host groups or get all host groups.
```
Comment (shekhar-j) [inline powermax/helper/port_helper.go:31]: Comment is for host group

File: powermax/provider/powermax.env [hunk 2]
```diff
@@ -1,6 +1,6 @@
 TF_ACC=1
-POWERMAX_ENDPOINT=https://<UNISPHERE_IP>:8443
-POWERMAX_USERNAME=
-POWERMAX_PASSWORD=
-POWERMAX_SERIAL_NUMBER=
+POWERMAX_ENDPOINT=http://localhost:3000
```
Comment (shekhar-j) [inline powermax/provider/powermax.env:2]: keep original,

**Review by shekhar-j — APPROVED** [review]

File: powermax/helper/port_helper.go (hunk 1 above)
Comment (doriac11) [reply powermax/helper/port_helper.go:31]: Done

File: powermax/provider/powermax.env (hunk 2 above)
Comment (doriac11) [reply powermax/provider/powermax.env]: yep sorry thought I reverted back.

**Review by nikitajoshi1 — APPROVED** [review]

File: examples/data-sources/powermax_port/variables.tf [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# variable "username" {
+#   type = string
+# }
+
+# variable "password" {
+#   type = string
+# }
+
```
Comment (nikitajoshi1) [inline examples/data-sources/powermax_port/variables.tf:8]: @doriac11 Please uncomment before merging the PR.

Comment (doriac11) [reply examples/data-sources/powermax_port/variables.tf:8]: Done, did it on the other PR forgot about this one.

**Pull request merged**